### PR TITLE
Feature/business mds refactor

### PR DESCRIPTION
**Pull request by bdmogal**
Summary:
1. Renamed Business* classes to just Metadata*
2. DefaultMetadataStore now serves two separate datasets based on the provided MetadataScope

This PR does not have any system metadata code. That will come in later.
This PR also does not include any upgrade steps. That will be a follow up PR.

It is split into 3 commits for convenience:
Commit1: Dataset renaming
Commit2: Refactoring to allow serving multiple datasets from DefaultMetadataStore
Commit3: Some missing Javadocs in MetadataAdmin and MetadataStore

Jira: [CDAP-4262](https://issues.cask.co/browse/CDAP-4262)
Build: http://builds.cask.co/browse/CDAP-DUT3169

### CODE REVIEW

**Comment by rohitsinha54** [comment]
Searching for Business in the project returns some results which will be better renamed appropriately.

**Comment by rohitsinha54** [comment]
Doing a search on Business on the whole project returns some results which should be renamed appropriately I think. Some of them you can do during the rebase too as they are in the http handler classes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 1]
```diff
@@ -61,18 +66,43 @@ void addProperties(Id.NamespacedId entityId, Map<String, String> properties)
   Set<MetadataRecord> getMetadata(Id.NamespacedId entityId) throws NotFoundException;
 
   /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: The doc here should say that the metadata will be of the given scope only

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 2]
```diff
@@ -61,18 +66,43 @@ void addProperties(Id.NamespacedId entityId, Map<String, String> properties)
   Set<MetadataRecord> getMetadata(Id.NamespacedId entityId) throws NotFoundException;
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: This function's doc need to updated to say that it will give metadata across scope

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 3]
```diff
@@ -62,81 +61,98 @@ public void addProperties(Id.NamespacedId entityId, Map<String, String> properti
...
 
   @Override
   public Map<String, String> getProperties(Id.NamespacedId entityId) throws NotFoundException {
     entityValidator.ensureEntityExists(entityId);
-    return businessMds.getProperties(entityId);
+    return metadataStore.getProperties(entityId);
+  }
+
+  @Override
+  public Map<String, String> getProperties(MetadataScope scope, Id.NamespacedId entityId) throws NotFoundException {
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: Same with this and the above which does not have scope

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 4]
```diff
@@ -62,81 +61,98 @@ public void addProperties(Id.NamespacedId entityId, Map<String, String> properti
...
 
   @Override
   public Set<String> getTags(Id.NamespacedId entityId) throws NotFoundException {
     entityValidator.ensureEntityExists(entityId);
-    return businessMds.getTags(entityId);
+    return metadataStore.getTags(entityId);
+  }
+
+  @Override
+  public Set<String> getTags(MetadataScope scope, Id.NamespacedId entityId) throws NotFoundException {
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: same

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 5]
```diff
@@ -61,18 +66,43 @@ void addProperties(Id.NamespacedId entityId, Map<String, String> properties)
...
    */
+  // TODO: This should perhaps return a Map<MetadataScope, Map<String, String>>
   Map<String, String> getProperties(Id.NamespacedId entityId) throws NotFoundException;
 
   /**
+   * @return a {@link Map} representing the business metadata of the specified {@link Id.Application},
+   * {@link Id.Program}, {@link Id.DatasetInstance} or {@link Id.Stream}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<String, String> getProperties(MetadataScope scope, Id.NamespacedId entityId) throws NotFoundException;
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Doc needs to be updated

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 6]
```diff
@@ -126,4 +156,18 @@ void addProperties(Id.NamespacedId entityId, Map<String, String> properties)
...
+   *
+   * @param namespaceId The namespace id to be filter the search by.
+   * @param searchQuery The query need to be executed for the search.
+   * @param type The particular type of CDAP object that the metadata need to be searched. If null all possible types
+   *             will be searched.
+   *
+   * @return a {@link Set} records for metadata search.
+   * @throws NotFoundException if there is not record found for particular query text.
+   */
+  Set<MetadataSearchResultRecord> searchMetadata(MetadataScope scope, String namespaceId, String searchQuery,
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Same

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 7]
```diff
@@ -164,9 +180,9 @@ public void validateTags(Id.NamespacedId entityId, String ... tags) throws Inval
 
   // Validate the key should not be be reserved {@link BusinessMetadataDataset.TAGS_KEY}
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: rename

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 8]
```diff
@@ -164,9 +180,9 @@ public void validateTags(Id.NamespacedId entityId, String ... tags) throws Inval
...
   private void validateTagReservedKey(Id.NamespacedId entityId, String key) throws InvalidMetadataException {
-    if (BusinessMetadataDataset.TAGS_KEY.equals(key.toLowerCase())) {
+    if (MetadataDataset.TAGS_KEY.equals(key.toLowerCase())) {
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: So these tag keys are like keywords for metadata and we didn't allow user's to set metadata with these key right ? 
Now since we are adding system metadata  and even though they will write to different tables but when a user does getTags() we return a set so does this keys needs to be updated to now allow system ones too ? or are  the keys uniquely identified by with their scope ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/MetadataStore.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+ * Defines operations on {@link MetadataDataset} for both system and user metadata.
+ *
+ * Operations supported for a specified {@link Id.NamespacedId}:
+ * <ul>
+ *   <li>Adding new metadata: Supported for a single scope.</li>
+ *   <li>Retrieving metadata: Supported for a specified scope as well as across both scopes.</li>
+ *   <li>Removing metadata: Supported for a specified scope as well as across both scopes.</li>
+ * </ul>
+ */
+public interface MetadataStore {
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/MetadataStore.java]: Lets says in function docs which does not have scopes that it will be scope for clarity.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 10]
```diff
@@ -61,18 +66,43 @@ void addProperties(Id.NamespacedId entityId, Map<String, String> properties)
...
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<String, String> getProperties(MetadataScope scope, Id.NamespacedId entityId) throws NotFoundException;
+
+  /**
+   * @return all the tags for the specified {@link Id.Application}, {@link Id.Program}, {@link Id.DatasetInstance} or
+   * {@link Id.Stream}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  // TODO: This should perhaps return a Map<MetadataScope, Iterable<String>>
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Yes I think it should too else if a user stores the same tag which we have in system will cause issues as its a Set

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java (hunk 5 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Same for this too right ? With scope else the key for user and system cannot be same

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java (hunk 10 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: This PR does a lot of refactoring already, I would prefer to do this in a separate PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: Good point regarding tags. What I'm unsure of at this point is how we will depict this in the UI. If the UI will do one call and expect a set of tags across SYSTEM and USER scopes, then I think we may need some special handling. Otherwise, we should just group the output of getAllTags by SCOPE, and let clients (UI, CLI) handle them separately.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: prefer to have all docs in the MetadataAdmin class

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: will mention it in the interface if it is missing.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java (hunk 5 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Hmmm, the user passes in scope, and this is a get API, so not sure what's the confusion here.

Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Oh my comment was on the Map<> returned by get without scope. The case is similar to tags right ? We should return with scope because if the key in the user and system metadata is same it will lead to issue.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java [hunk 11]
```diff
@@ -76,7 +77,8 @@ public MetadataHttpHandler(MetadataAdmin metadataAdmin) {
   public void getAppMetadata(HttpRequest request, HttpResponder responder,
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: So all the handlers make request with scope user. How can an user retrieve system metadata then if they want to see it ? o

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: I'll add support for a 'scope' query param in a later PR. This PR is parity with current functionality + ability to create a system metadata dataset only.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 12]
```diff
@@ -400,9 +398,9 @@ private MetadataRecord getSnapshotBeforeTime(Id.NamespacedId targetId, long time
   }
 
   // Helper method to execute IndexedTable search on target index column.
-  List<BusinessMetadataRecord> executeSearchOnColumns(final String namespaceId, final String column,
+  List<MetadataEntry> executeSearchOnColumns(final String namespaceId, final String column,
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: private

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java (hunk 5 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: But users have already specified the scope in this API right? Why should it be part of the return value too?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 13]
```diff
@@ -124,9 +124,8 @@ public void setProperty(Id.NamespacedId targetId, String key, String value) {
    * @param tags the tags to set
    */
   private void setTags(Id.NamespacedId targetId, String ... tags) {
-    BusinessMetadataRecord tagsRecord =
-      new BusinessMetadataRecord(targetId, TAGS_KEY, Joiner.on(TAGS_SEPARATOR).join(tags));
-    setBusinessMetadata(tagsRecord, MetadataType.TAG);
+    MetadataEntry tagsEntry = new MetadataEntry(targetId, TAGS_KEY, Joiner.on(TAGS_SEPARATOR).join(tags));
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Not related to this PR but a lot of these static of this class can be static. Lets change them.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetDefinition.java [hunk 14]
```diff
@@ -31,15 +31,15 @@
 /**
  * Define the Dataset for metadata.
  */
-public class BusinessMetadataDefinition extends AbstractDatasetDefinition<BusinessMetadataDataset, DatasetAdmin> {
+public class MetadataDatasetDefinition extends AbstractDatasetDefinition<MetadataDataset, DatasetAdmin> {
 
   public static final String METADATA_INDEX_TABLE_NAME = "metadata_index";
-  public static final String INDEXED_COLS = BusinessMetadataDataset.KEYVALUE_COLUMN + "," +
-    BusinessMetadataDataset.CASE_INSENSITIVE_VALUE_COLUMN;
+  public static final String INDEXED_COLS = MetadataDataset.KEYVALUE_COLUMN + "," +
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetDefinition.java]: change to private and the one above too

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java [hunk 15]
```diff
@@ -21,20 +21,21 @@
 import javax.annotation.Nullable;
 
 /**
- * Represent Busineess Metadata for CDAP.
+ * Represents a single Metadata entry for a CDAP Entity.
  */
-public class BusinessMetadataRecord {
+public class MetadataEntry {
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java]: Should me  also store the scope of entry here ?

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java]: Nope, this is for an entry in the table. The table is already scoped.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 13 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: done

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/MetadataStore.java (hunk 9 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/MetadataStore.java]: went through all javadocs and updated them to current state. please check again.

**Comment by bdmogal** [comment]
@rohitsinha54 addressed comments. Please review again.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DefaultMetadataStore.java [hunk 16]
```diff
@@ -425,28 +479,35 @@ private void publish(MetadataRecord previous, MetadataRecord additions, Metadata
...
-        dsFramework, BUSINESS_METADATA_INSTANCE_ID, BusinessMetadataDataset.class.getName(),
+        dsFramework, getMetadataDatasetInstance(scope), MetadataDataset.class.getName(),
         DatasetProperties.EMPTY, DatasetDefinition.NO_ARGUMENTS, null);
     } catch (Exception e) {
       throw Throwables.propagate(e);
     }
   }
 
+  private Id.DatasetInstance getMetadataDatasetInstance(MetadataScope scope) {
+    if (MetadataScope.USER == scope) {
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DefaultMetadataStore.java]: nit:     
```
return MetadataScope.USER == scope ? BUSINESS_METADATA_INSTANCE_ID : SYSTEM_METADATA_INSTANCE_ID;
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java (hunk 8 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: Okay. Lets revisit it. Do you think we need a jira for this ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 11 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Should we open a jira to support retrieval of system metadata once we start adding system metadata ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java]: Hmmm, not sure what that jira will describe. I think we can handle it during the implementation of system metadata. Or do you think a jira with the title "Response of listing all properties or all tags for an entity should make it easy to distinguish between system and user metadata" makes sense?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 11 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: https://issues.cask.co/browse/CDAP-4264 and its subtasks should cover it?

**Comment by rohitsinha54** [comment]
One minor nit besides thats LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 11 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: cool

**Comment by bdmogal** [comment]
Thanks @rohitsinha54. Will squash and merge after build passes.

**Comment by bdmogal** [comment]
Build passed - http://builds.cask.co/browse/CDAP-DUT3169-11. Merging.
